### PR TITLE
samples: net: mqtt: Add a sample to connect to AWS IoT core

### LINKED ISSUE #22866: Amazon Web Services sample
Hi ,
I dont see an example related to publishing our data to Amazon Web Services
https://docs.zephyrproject.org/latest/samples/index.html

If we have a sample it will be really helpful

### PR DESCRIPTION
**Pull request by lucasdietrich**
I would like to propose the addition of a new sample application to Zephyr RTOS, which implements an MQTT client that can publish messages to AWS IoT Core using the MQTT protocol. The sample showcases the implementation of key features such as establishing a TLS 1.2 connection with AWS IoT Core, subscribing to a topic, and publishing data. Additionally, this sample is designed to pass the test suite for the AWS Device Qualification Program (DQP).

This sample can be used as a starting point for developers who are interested in building their own IoT solutions using Zephyr RTOS and AWS IoT Core.

The sample has been tested on the ST NUCLEO-F429ZI board and QEMU x86. I have taken other net/cloud samples as examples, so there are similarities in parts of the code.

Please let me know if you have any questions, feedbacks or concerns.

Fixes #22866

### CODE REVIEW

File: samples/net/cloud/aws_iot_mqtt/Kconfig [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+config AWS_TEST_SUITE_NONE
+	bool "No test suite running"
+config AWS_TEST_SUITE_DQP
+	bool "Device Qualification Program (DQP)"
+	help
+		Make sure your region supports AWS Device Advisor for DQP
+config AWS_TEST_SUITE_RECV_QOS1
+	bool "Test suite for receiving QoS 1 messages"
+	help
+		For single test case "MQTT Client Puack QoS1"
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/Kconfig:49]: I'd still add an empty line in between choice options, currently, it's a bit hard to read (at least that was my first impression).

File: samples/net/cloud/aws_iot_mqtt/src/dhcp.h [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __DHCP_H__
+#define __DHCP_H__
+
+void app_dhcpv4_startup(void);
+
+#endif /* not __DHCP_H__ */
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/dhcp.h:14]: Why `not`?

File: samples/net/cloud/aws_iot_mqtt/src/dhcp.h [hunk 3]
```diff
@@ -0,0 +1,14 @@
+/* DHCPv4 client startup. */
+
+/*
+ * Copyright (c) 2018 Linaro Ltd
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/dhcp.h:4]: The copyright seems wrong.

File: samples/net/cloud/aws_iot_mqtt/src/tls_config/user-tls.h [hunk 4]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) 2023 Lucas Dietrich <ld.adecy@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define MBEDTLS_AES_ROM_TABLES
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/tls_config/user-tls.h:7]: There's `CONFIG_MBEDTLS_AES_ROM_TABLES` for that.

File: samples/net/cloud/aws_iot_mqtt/src/tls_config/user-tls.h [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+#define MBEDTLS_HAVE_TIME
+#define MBEDTLS_HAVE_TIME_DATE
+#define MBEDTLS_PLATFORM_TIME_ALT
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/tls_config/user-tls.h:11]: Is this really needed? The sample doesn't seem to be using `mbedtls_platform_set_time()`

File: samples/net/cloud/aws_iot_mqtt/src/tls_config/user-tls.h [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+#undef MBEDTLS_SSL_KEEP_PEER_CERTIFICATE
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/tls_config/user-tls.h:13]: Why is this needed? The generic mbedTLS config file from Zephyr does not set this config.

File: samples/net/cloud/aws_iot_mqtt/src/dhcp.c [hunk 7]
```diff
@@ -0,0 +1,66 @@
+/* DHCPv4 client startup. */
+
+/*
+ * Copyright (c) 2018 Linaro Ltd
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/dhcp.c:4]: The copyright seems wrong.

File: samples/net/cloud/aws_iot_mqtt/src/main.c [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+#include <zephyr/net/sntp.h>
+#include <zephyr/net/tls_credentials.h>
+#include <zephyr/random/rand32.h>
+
+
+#if defined(CONFIG_MBEDTLS_MEMORY_DEBUG)
+#include <mbedtls/memory_buffer_alloc.h>
+#endif
+
+#include <sys/socket.h>
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/main.c:30]: It seems to me you should rather include `#include <zephyr/net/socket.h>` given that you use `zsock_*` APIs, not POSIX header. Are posix APIs even needed for this sample?

File: samples/net/cloud/aws_iot_mqtt/src/main.c [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+
+	rc = sntp_simple(SNTP_SERVER, SYS_FOREVER_MS, &now);
+	if (rc == 0) {
+		tspec.tv_sec = now.seconds;
+		tspec.tv_nsec = ((uint64_t)now.fraction * (1000lu * 1000lu * 1000lu)) >> 32;
+
+		clock_settime(CLOCK_REALTIME, &tspec);
+
+		LOG_DBG("Acquired time from NTP server: %u", (uint32_t)tspec.tv_sec);
+
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/main.c:428]: Unneeded empty line.

File: samples/net/cloud/aws_iot_mqtt/src/main.c [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+	struct zsock_addrinfo *ai = NULL;
+
+	const struct zsock_addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM,
+		.ai_protocol = 0,
+	};
+
+	ret = zsock_getaddrinfo(CONFIG_AWS_ENDPOINT, AWS_BROKER_PORT, &hints, &ai);
+
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/main.c:447]: No need for an empty line here.

File: samples/net/cloud/aws_iot_mqtt/src/main.c [hunk 11]
```diff
@@ -0,0 +1,494 @@
...
+	if (ret == 0) {
+		memcpy(broker, ai->ai_addr, MIN(ai->ai_addrlen, sizeof(struct sockaddr_storage)));
+
+		char addr_str[INET_ADDRSTRLEN];
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/main.c:451]: It's more of a convention, but we aim to define variables at the top of the block.

File: samples/net/cloud/aws_iot_mqtt/src/main.c [hunk 12]
```diff
@@ -0,0 +1,494 @@
...
+static uint32_t messages_received_counter;
+static bool do_publish;	  /* Trigger client to publish */
+static bool do_subscribe; /* Trigger client to subscribe */
+
+#define TLS_TAG_DEVICE_CERTIFICATE 1
+#define TLS_TAG_DEVICE_PRIVATE_KEY 1
+#define TLS_TAG_AWS_CA_CERTIFICATE 1
+
+static sec_tag_t sec_tls_tags[] = {
+	TLS_TAG_DEVICE_CERTIFICATE,
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/main.c:64]: Ah, I've started to wonder how does it work if you only provide device certificate tag and no root CA, but now I see they're all regsitered under the same tag number.

I think it'd be a bit more readable, if root CA received its own tag, as it's not related to the device credentials. Device private tag and cert should share the same tag though, as they come in couple.

File: samples/net/cloud/aws_iot_mqtt/src/main.c [hunk 13]
```diff
@@ -0,0 +1,494 @@
...
+	}
+
+exit:
+	return ret;
+}
+
+static int subscribe_topic(void)
+{
+	int ret;
+
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/main.c:99]: Those empty lines between variables defined are not needed.

File: samples/net/cloud/aws_iot_mqtt/src/main.c [hunk 14]
```diff
@@ -0,0 +1,494 @@
...
+	aws_client_setup();
+
+	rc = aws_client_try_connect();
+	if (rc != 0) {
+		goto cleanup;
+	}
+
+	fds.fd = client_ctx.transport.tcp.sock;
+	fds.events = POLLIN | POLLHUP | POLLERR;
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/main.c:372]: POLLHUP and POLLERR are `revents` only - no need to register for them.

File: samples/net/cloud/aws_iot_mqtt/src/main.c [hunk 15]
```diff
@@ -0,0 +1,494 @@
...
+		timeout = mqtt_keepalive_time_left(&client_ctx);
+		rc = poll(&fds, 1u, timeout);
+		if (rc >= 0) {
+			if (fds.revents & POLLIN) {
+				rc = mqtt_input(&client_ctx);
+				if (rc != 0) {
+					LOG_ERR("Failed to read MQTT input: %d", rc);
+					break;
+				}
+			}
```
Comment (rlubos) [inline samples/net/cloud/aws_iot_mqtt/src/main.c:384]: I'd also check for POLLHUP and POLLERR here, and break if set.

**Review by rlubos — COMMENTED** [review]
Thanks @lucasdietrich for the effort in contributing the sample. Looks pretty neat.

File: samples/net/cloud/aws_iot_mqtt/src/dhcp.c [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+{
+	int i;
+	bool notified = false;
+
+	if (mgmt_event != NET_EVENT_IPV4_ADDR_ADD) {
+		return;
+	}
+
+	for (i = 0; i < NET_IF_MAX_IPV4_ADDR; i++) {
+
```
[the remark this answers is not in the document]
Comment (lucasdietrich) [reply samples/net/cloud/aws_iot_mqtt/src/dhcp.c:37]: See my comment on dhcp.c

File: samples/net/cloud/aws_iot_mqtt/src/main.c (hunk 13 above)
Comment (lucasdietrich) [reply samples/net/cloud/aws_iot_mqtt/src/main.c]: Fixed, do you identify these style inconsistencies on your own or do you have a script for it?

File: samples/net/cloud/aws_iot_mqtt/src/main.c (hunk 14 above)
Comment (lucasdietrich) [reply samples/net/cloud/aws_iot_mqtt/src/main.c]: Didn't know, fixed.

File: samples/net/cloud/aws_iot_mqtt/src/main.c (hunk 8 above)
Comment (lucasdietrich) [reply samples/net/cloud/aws_iot_mqtt/src/main.c]: I didn't made the change yet, do you suggest to remove `#include <poll.h>` and `#include <unistd.h>`, then use `zsock_*` variants for poll related stuff ? It that case, we could almost disable `CONFIG_POSIX_API` (last POSIX function used is `clock_settime` from  `#include <time.h>`).

If we decide to pursue this approach, my suggestion would be to enable `CONFIG_POSIX_CLOCK` to still have the convenient function `clock_settime` through `#include <zephyr/posix/time.h>` rather than `#include <time.h>`.

What are your thoughts on this?

File: samples/net/cloud/aws_iot_mqtt/src/dhcp.h (hunk 3 above)
Comment (rlubos) [reply samples/net/cloud/aws_iot_mqtt/src/dhcp.h:4]: Sounds fine.

File: samples/net/cloud/aws_iot_mqtt/src/main.c (hunk 13 above)
Comment (rlubos) [reply samples/net/cloud/aws_iot_mqtt/src/main.c]: I'm afraid this is not verified by the scripts we have, so yes, manual spot.

File: samples/net/cloud/aws_iot_mqtt/src/main.c (hunk 8 above)
Comment (rlubos) [reply samples/net/cloud/aws_iot_mqtt/src/main.c]: I'd just prefer to be consistent with the API usage, if started to use `zsock_*` APIs, just stick to it.

As for clock, enabling `CONFIG_POSIX_CLOCK` only seems fine.

File: samples/net/cloud/aws_iot_mqtt/src/dhcp.h (hunk 3 above)
Comment (lucasdietrich) [reply samples/net/cloud/aws_iot_mqtt/src/dhcp.h:4]: I made the changes and updated the license in `dhcp.h` and `dhcp.c`

**Review by rlubos — APPROVED** [review]
LGTM

**Review by jukkar — APPROVED** [review]
